### PR TITLE
Update Compiler.php

### PR DESCRIPTION
**Pull request by ghost**
line 22: private $lastLine; // It can be safely removed, see also line 52 & 175

line 51 $this->lastLine = null; // deletet line

line 175: if ($node->getTemplateLine() != $this->lastLine) { // set it from lastLine to sourceLine

line 175: if ($node->getTemplateLine() !== $this->lastLine) { // conditional can be simplified for strict comparison

line 189: return $this->debugInfo ?? []; // handle the case when no template lines have been added and returning an empty array in that case

insert line to 209: $this->indentation ??= 0;  // check to prevent the indentation from becoming negative

### CODE REVIEW

File: src/Compiler.php [hunk 1]
```diff
@@ -48,16 +45,13 @@ public function getSource(): string
...
-        $this->lastLine = null;
         $this->source = '';
         $this->debugInfo = [];
         $this->sourceOffset = 0;
         // source code starts at 1 (as we then increment it when we encounter new lines)
         $this->sourceLine = 1;
         $this->indentation = $indentation;
         $this->varNameSalt = 0;
-
-        return $this;
```
Comment (GromNaN) [inline src/Compiler.php:60]: This is a breaking change. The phpdoc tell the returned value.

File: src/Compiler.php [hunk 2]
```diff
@@ -186,7 +180,7 @@ public function getDebugInfo(): array
     {
         ksort($this->debugInfo);
 
-        return $this->debugInfo;
+        return $this->debugInfo ?? [];
```
Comment (GromNaN) [inline src/Compiler.php:182]: If debugInfo is not an array, the call to ksort would fail.

File: src/Compiler.php [hunk 3]
```diff
@@ -206,6 +200,7 @@ public function indent(int $step = 1)
      */
     public function outdent(int $step = 1)
     {
+        $this->indentation ??= 0;
```
Comment (GromNaN) [inline src/Compiler.php:202]: Not necessary, the comparison with null is fine.

**Review by GromNaN — CHANGES_REQUESTED** [review]
Did you use any static analysis tool to detect this changes? I'm not sure they are correct, except the property promotion which is syntactic sugar.

Maybe add a test if you found edge cases.

File: src/Compiler.php [hunk 4]
```diff
@@ -169,7 +163,7 @@ public function repr($value)
      */
     public function addDebugInfo(Node $node)
     {
-        if ($node->getTemplateLine() != $this->lastLine) {
+        if ($node->getTemplateLine() !== $this->sourceLine) {
```
Comment (GromNaN) [inline src/Compiler.php]: `$this->lastLine` is still used line 173.

File: src/Environment.php [hunk 5]
```diff
@@ -512,19 +512,33 @@ public function compile(Node $node): string
...
-        try {
-            return $this->compile($this->parse($this->tokenize($source)));
-        } catch (Error $e) {
-            $e->setSourceContext($source);
-            throw $e;
-        } catch (\Exception $e) {
-            throw new SyntaxError(sprintf('An exception has been thrown during the compilation of a template ("%s").', $e->getMessage()), -1, $source, $e);
-        }
-    }
+	public function compileSource(Source $source): string
```
Comment (GromNaN) [inline src/Environment.php]: Indenting with tabs is incorrect.

**Review by GromNaN — CHANGES_REQUESTED** [review]

File: src/Environment.php (hunk 5 above)
Comment (ghost) [reply src/Environment.php]: I have rewritten it as follows:

public function compileSource(Source $source): string
{
    	$cacheKey = $this->getCacheKey($source->getName());
    	$content = $this->cache->get($cacheKey);

    	if ($content === null) {
        	$content = $this->compile($this->parse($this->tokenize($source)));
        	$this->cache->put($cacheKey, $content);
    	}

    	return $content;
}

public function tokenize(Source $source): array
{
    	return $source->getTokens();
}

File: src/Compiler.php (hunk 4 above)
Comment (ghost) [reply src/Compiler.php]: rewritten it as follows:

public function addDebugInfo(Node $node)
{
    	$this->sourceLine = $node->getTemplateLine();

    	$this->write("// line {$node->getTemplateLine()}\n");

    	$templateLines = substr_count($node->getSource(), PHP_EOL);
    	$this->sourceLine += $templateLines;
    	$this->sourceOffset = strlen($this->source);

    	$this->source .= $node->getSource();

    	$this->debugInfo[$node->getTemplateLine()] = $node->getTemplateLine();

    	return $this;
}

File: src/Compiler.php (hunk 2 above)
Comment (ghost) [reply src/Compiler.php:182]: Do we have data in debuginfo that is not available as an array, although the function :: requires an array ?

File: src/Compiler.php (hunk 3 above)
Comment (ghost) [reply src/Compiler.php:202]: As it was originally set with step=1, 0 is also set as an error. Should this be explicit?

File: src/Environment.php (hunk 5 above)
Comment (GromNaN) [reply src/Environment.php]: Use php-cs-fixer to fix the coding style before committing changes, please.

File: src/Compiler.php [hunk 6]
```diff
@@ -218,6 +227,6 @@ public function outdent(int $step = 1)
 
     public function getVarName(): string
     {
-        return sprintf('__internal_compile_%d', $this->varNameSalt++);
+        return \sprintf('__internal_compile_%d', $this->varNameSalt++);
```
Comment (GromNaN) [inline src/Compiler.php:215]: The root namespace is added only for functions that have a special opcode compilation. This is done by php-cs-fixer. 
https://cs.symfony.com/doc/rules/function_notation/native_function_invocation.html

The function sprintf don't need to be root-namespaced.

File: src/Compiler.php (hunk 2 above)
Comment (GromNaN) [reply src/Compiler.php:182]: debugInfo is always an array.

File: src/Environment.php [hunk 7]
```diff
@@ -40,7 +40,7 @@
  */
 class Environment
 {
-    public const VERSION = '4.0.0-DEV';
+    public const VERSION = '4.0.0-DEV-modified-by-owsPro';
```
Comment (GromNaN) [inline src/Environment.php]: To be reverted.

**Review by GromNaN — CHANGES_REQUESTED** [review]
You are doing too many unrelated changes in this PR, it's not going to be something that can be merged. 
Please explain your goal so that we can help having a clean targeted diff.

File: src/Compiler.php [hunk 8]
```diff
@@ -147,11 +141,11 @@ public function repr($value)
         } elseif (\is_array($value)) {
             $this->raw('array(');
             $first = true;
-            foreach ($value as $key => $v) {
+            \foreach ($value as $key => $v) {
```
Comment (smnandre) [inline src/Compiler.php]: Is phpcsfixer doing this ? oO 

I've never seen `\foreach` before

File: src/Compiler.php [hunk 9]
```diff
@@ -167,27 +161,41 @@ public function repr($value)
...
     public function getDebugInfo(): array
     {
-        ksort($this->debugInfo);
+        \ksort($this->debugInfo);
 
-        return $this->debugInfo;
+        return $this->debugInfo ?? [];
     }
+    
+    function sortData($data)
```
Comment (smnandre) [inline src/Compiler.php]: Where is it used ?

File: src/Environment.php [hunk 10]
```diff
@@ -798,9 +804,7 @@ public function mergeGlobals(array $context): array
         // we don't use array_merge as the context being generally
         // bigger than globals, this code is faster.
         foreach ($this->getGlobals() as $key => $value) {
-            if (!\array_key_exists($key, $context)) {
-                $context[$key] = $value;
-            }
+            $context[$key] = $value;
```
Comment (smnandre) [inline src/Environment.php:807]: I think that would be a big BC break, no ?

File: src/Compiler.php (hunk 8 above)
Comment (ghost) [reply src/Compiler.php]: Nope, is corrected!

File: src/Environment.php [hunk 11]
```diff
@@ -136,7 +136,7 @@ public function enableDebug()
     /**
      * Disables debugging mode.
      */
-    public function disableDebug()
+    public function disableDebug(): void
```
Comment (stof) [inline src/Environment.php]: Adding those return types should first be done in 3.x by adding `@return` annotations to warn child classes that they need to add it. Otherwise, we don't have the proper migration path.

**Review by stof — COMMENTED** [review]
With this PR that changes the formatting of the code not respecting the coding standards of the project, it is impossible to properly review it.
Thus, it does not even clearly explain its goal.

In its current state, my vote is to reject it.

File: src/Compiler.php (hunk 9 above)
Comment (ghost) [reply src/Compiler.php]: Has been removed again and should not appear at all!

File: src/Environment.php (hunk 11 above)
Comment (ghost) [reply src/Environment.php]: Ok!

File: src/Environment.php (hunk 5 above)
Comment (ghost) [reply src/Environment.php]: I have it now.

File: src/Environment.php (hunk 7 above)
Comment (ghost) [reply src/Environment.php]: It is now.

File: src/Environment.php [hunk 12]
```diff
@@ -514,17 +514,22 @@ public function compile(Node $node): string
...
+        $content = $this->cache->get($cacheKey);
+
+        if ($content === null) {
+            $content = $this->compile($this->parse($this->tokenize($source)));
+            $this->cache->put($cacheKey, $content);
         }
+
+        return $content;
+    }
+    public function tokenize(Source $source): array
```
Comment (smnandre) [inline src/Environment.php:527]: This looks like a c/c mistake ..  ?

File: src/Compiler.php (hunk 6 above)
Comment (ghost) [reply src/Compiler.php:215]: Rule native_function_invocation¶
Adds a leading \ before the function call to speed up resolution.

Warning¶
Using this rule is risky¶
Risky if one of the functions is overridden.

Well, then you should never use rule native_function_invocation, as this could always be risky depending on the context, right?

Comment (xabbuh) [reply src/Compiler.php:215]: The point is not about the rule being risky but about the fact that the rule as applied here only prefixes some functions that can be optimized by the compiler with a backslash.

**Comment by fabpot** [comment]
@owsPro I don't see what you're trying to do here. Can you clarify your goal before investing more time? As of now, I think it's better to just close.

File: src/Compiler.php (hunk 6 above)
Comment (ghost) [reply src/Compiler.php:215]: This is not objected to by the php cs fixer. This link https://stackoverflow.com/questions/55419673/php7-adding-a-slash-to-all-standard-php-functions-php-cs-fixer-rule shows that it is generally not that simple.

**Comment by ghost** [comment]
If fabpot comes to the realisation that it is better to close here, it will probably be a wasted effort.

**Comment by stof** [comment]
Well, this is the same than my comment: you don't explain what you are trying to do. That's the most important part when starting a PR.
